docs/design.md

```diff
@@ -0,0 +1,279 @@
+
+In order not to sync the whole chain of advertisements in every request and exclusively fetching the missing range, 
+the indexer node will have to be able to flag in the Graphsync request a "termination link",
+i.e. the link from which Graphsync can stop traversing the chain and stop recursing links.
+
+Once the sync process is finished, the indexer node updates the latest AdvertisementID seen for provider to `latest`.
+
+```go
+// Example of internal data structure of an index of data provider.
+type AdvertisementID Index_Link
+
+type Index struct {
+        Previous: Index_Link{},
+        Entries: []Entry{
+                IsRm:           bool, 
+                Cids:           []cid.Cid,        // NOTE: This can also be string.
+                Metadata:       []byte,
+        },
+} 
+```
+
+_NOTE: We need to build the selector or the IPLD storer in the indexer node so that it checks the latest Advertisement
+seen and it terminates Graphsync's DAG traversal conveniently._
+
+### IPLD-unaware Ingestion 
+> Out of scope of MVP. For the MVP we are only considering IPLD-aware ingestion, and we enforce the use of IPLD to providers.
+
+It implements a libp2p request-response ingestion protocol that indexer nodes can use when interacting with providers that are not indexing their
+data using IPLD, i.e. Advertisements are not directly linked to data. This protocol is inspired by 
+[Filecoin's Block Sync protocol](https://spec.filecoin.io/#section-algorithms.block_sync). The underlaying idea is the same, to "sync a chain of data".
+
+Thus, when calling `Sync(ctx, p, latest)`, indexer node sends a first request for ingestion as:
+```go
+req = IngestionRequest{
+        ID: latest,
+        Start: 0, 
+}
+```
+
+Provider will answer with the list of entries for `latest` pointing to the previous `AdvertisementID` in the `Advertisement` field. `Size` specifies
+the total number of entries that the advertisement include to give the indexer knowledge of when all entries for the update have been received.
+
+```go
+resp = IngestionResponse {
+        Size: 2
+        Error: 0
+        Start: 0
+        Advertisement: Advertisement{
+                ID: latest,
+                Previous: "latest-1",
+                IndexerID: Link_Index,
+                Provider: p,
+                Signature: []byte{},  // The original signature of the advertisement.
+                GraphSupport: false, 
+        }
+        Entries: {
+                []Entry{
+                        Cids: {Cid1, Cid2, ...},
+                        Metadata: []byte{},
```

Comment:
Look at the [design doc](https://www.notion.so/protocollabs/Indexer-Node-Design-4fb94471b6be4352b6849dc9b9527825#a9b0aa68c1374ab4897aac7917e7262f) to make sure these align.

Reply:
A few thoughts regarding `Entries` and and `Requests`:
- Do we really need to add `prev_index` in requests as described in the design doc? If we are building a message-oriented protocol and we want the server side to be stateless, it is better to tailor requests of indices as we need them. For Graphsync, we will just send a selector for the latest advertisement and terminate when we reach our current state.

- Regarding Entries, do we expect `Cids` and `RmCids` to be set in the same update? For Filecoin miners I don't think so, but for other use cases probably yes. I'll update it accordingly.

Reply:
The indexer looks at the previous entry to see if already has the previous change set, or if it needs to fetch that first (since differences are applied in order).  Without `prev_index` the indexer will not know whether it already has the previous index or whether it missed some updates and needs to fetch them.  When getting previous indexes, the indexer knows when to stop when it sees a `prev_index` that it already has.

> For Graphsync, we will just send a selector for the latest advertisement and terminate when we reach our current state.

How does the indexer know if the selector only needs the latest, or if it needs the last 3 because some were missed?

CIDs to add and CIDs to remove can be in the same entry or in different entries.  It is up to the provider how they want to break it up.  I can see reasons why a provider might want to do one or the other (depending on how their diff calculation works), so I figured we should allow both.
Maybe I am not understanding how diffs work with IPLD, but to get the correct set of diffs as described in the design doc,  the advertisement will need to include the previous index ID.

Reply:
@hannahhoward suggested it would not be hard to add an extension to graphsync to include a 'skip subdag' extension to allow the indexer in its request to provide its most recent synced link as an indication that it has already received objects from that point back and to stop traversal at that link.

Reply:
> The indexer looks at the previous entry to see if already has the previous change set, or if it needs to fetch that first (since differences are applied in order). Without prev_index the indexer will not know whether it already has the previous index or whether it missed some updates and needs to fetch them.

Of course, advertisements should include a pointer to the previous advertisement in the chain. What I meant is that for ingestion requests, there is no need for `prev_index` (as suggested in the design doc) as I think the one keeping the state of what to request next should be the indexer and not the provider.

From what I understand:
- Providers are responsible for keeping a local index of the content they are providing, and linking updates over the index to advertisements. Advertisements are linked sequentially to each other building a chain. The provider is the one responsible for linking Advertisement to the set of new and removed CIDs in their index however it wants (storing a static IPLD structure, computing this relationship in every request, etc.). The only requirements is that advertisements should be linked to entries in a way that they can be authenticated when ingested by the indexer.
- Indexers receiving an unseen advertisement are responsible for following the links and gathering the updates up to their last sync. `latest_ad --> entries --> previous_ad --> entries --> ... --> last_sync`.
- If the provider supports graphsync it should build its internal data structures and the links between advertisement and entries using a DAG structure so Graphsync can seamlessly do the traversal from advertisement to linked entries, to previous advertisement, to linked entries, and so on. For the non-graphsync case, the indexer will perform requests for each advertisement ID and proactively traverse the chain by making stateless requests to the provider for entries related to an advertisement until it fully syncs.

Adding a quick figure to support my explanation (in case it helps): 
![image](https://user-images.githubusercontent.com/6717133/127449690-2ca0975c-8dd0-42f9-b12a-6bf5f0bfd40c.png)

Does this make sense to you or am I missing something?

Reply:
Ok, I understand what you are saying.  Appreciate explanation.  Diagram is great to include in this doc aslo.